File: python/src/activities.py
import asyncio
import logging

from temporalio import activity
from temporalio.exceptions import ApplicationError

from shared import BookTripInput

logging.basicConfig(level=logging.INFO)


@activity.defn
async def book_flight(input: BookTripInput) -> str:
    activity.logger.info(f"Booking flight: {input.flightId}")

    await asyncio.sleep(1)

    if "flaky" in input.flightId.lower():
        # a transient error, which will be retried
        if activity.info().attempt < 6:
            raise ApplicationError("Flight booking service is currently unavailable")

    return f"Booked flight: {input.flightId}"


@activity.defn
async def book_hotel(input: BookTripInput) -> str:
    activity.logger.info(f"Booking hotel: {input.hotelId}")

    await asyncio.sleep(1)

    if "buggy" in input.hotelId.lower():
        # a simulated bug
        error = True
        if error:
            raise Exception("Error due to bug in code")

    return f"Booked hotel: {input.hotelId}"


@activity.defn
async def book_car(input: BookTripInput) -> str:
    activity.logger.info(f"Booking car {input.carId}")

    await asyncio.sleep(1)

    if "invalid" in input.carId.lower():
        # a business error, which cannot be retried
        raise ApplicationError(f"Car {input.carId} is invalid", type="InvalidCar", non_retryable=True)

    return f"Booked car: {input.carId}"


@activity.defn
async def notify_user(input: BookTripInput) -> str:
    activity.logger.info(f"Notifying user: {input.userId}")

    await asyncio.sleep(1)

    return f"Notified user: {input.userId}"


@activity.defn
async def undo_book_flight(input: BookTripInput) -> str:
    activity.logger.info(f"Undo flight booking: {input.flightId}")

    await asyncio.sleep(1)

    return f"Unbooked flight: {input.flightId}"


@activity.defn
async def undo_book_hotel(input: BookTripInput) -> str:
    activity.logger.info(f"Undo hotel booking: {input.hotelId}")

    await asyncio.sleep(1)

    return f"Unbooked hotel: {input.hotelId}"


@activity.defn
async def undo_book_car(input: BookTripInput) -> str:
    activity.logger.info(f"Undo car booking: {input.carId}")

    await asyncio.sleep(1)

    return f"Unbooked car: {input.carId}"

File: dotnet/src/Activities.cs
using Microsoft.Extensions.Logging;
using Temporalio.Activities;
using Temporalio.Exceptions;

namespace TemporalTripBooking;

public class TripActivities
{
    [Activity]
    public static async Task<string> BookFlight(BookTripInput input)
    {
        var logger = ActivityExecutionContext.Current.Logger;
        logger.LogInformation("Booking flight: {FlightId}", input.FlightId);

        await Task.Delay(TimeSpan.FromSeconds(1));

        if (input.FlightId.Contains("flaky", StringComparison.OrdinalIgnoreCase) &&
            ActivityExecutionContext.Current.Info.Attempt < 6)
            throw new Exception("Flight booking service is currently unavailable");

        return $"Booked flight: {input.FlightId}";
    }

    [Activity]
    public static async Task<string> BookHotel(BookTripInput input)
    {
        var logger = ActivityExecutionContext.Current.Logger;
        logger.LogInformation("Booking hotel: {HotelId}", input.HotelId);

        await Task.Delay(TimeSpan.FromSeconds(1));

        if (input.HotelId.Contains("buggy", StringComparison.OrdinalIgnoreCase))
        {
            var error = true;
            if (error) throw new Exception("Error due to bug in code");
        }

        return $"Booked hotel: {input.HotelId}";
    }

    [Activity]
    public static async Task<string> BookCar(BookTripInput input)
    {
        var logger = ActivityExecutionContext.Current.Logger;
        logger.LogInformation("Booking car: {CarId}", input.CarId);

        await Task.Delay(TimeSpan.FromSeconds(1));

        if (input.CarId.Contains("invalid", StringComparison.OrdinalIgnoreCase))
            throw new ApplicationFailureException(
                $"Car {input.CarId} is invalid", "InvalidCar", true);

        return $"Booked car: {input.CarId}";
    }

    [Activity]
    public static async Task<string> NotifyUser(BookTripInput input)
    {
        var logger = ActivityExecutionContext.Current.Logger;
        logger.LogInformation("Notifying user: {UserId}", input.UserId);

        await Task.Delay(TimeSpan.FromSeconds(1));

        return $"Notified user: {input.UserId}";
    }

    [Activity]
    public static async Task<string> UndoBookFlight(BookTripInput input)
    {
        var logger = ActivityExecutionContext.Current.Logger;
        logger.LogInformation("Undo flight booking: {FlightId}", input.FlightId);

        await Task.Delay(TimeSpan.FromSeconds(1));

        return $"Unbooked flight: {input.FlightId}";
    }

    [Activity]
    public static async Task<string> UndoBookHotel(BookTripInput input)
    {
        var logger = ActivityExecutionContext.Current.Logger;
        logger.LogInformation("Undo hotel booking: {HotelId}", input.HotelId);

        await Task.Delay(TimeSpan.FromSeconds(1));

        return $"Unbooked hotel: {input.HotelId}";
    }

    [Activity]
    public static async Task<string> UndoBookCar(BookTripInput input)
    {
        var logger = ActivityExecutionContext.Current.Logger;
        logger.LogInformation("Undo car booking: {CarId}", input.CarId);

        await Task.Delay(TimeSpan.FromSeconds(1));

        return $"Unbooked car: {input.CarId}";
    }
}